codegen/gateway.go

```diff
@@ -420,7 +420,71 @@ func NewEndpointSpec(
 		ClientMethod:       clientMethod,
 	}
 
-	return augmentEndpointSpec(espec, endpointConfigObj, midSpecs)
+	mandatoryMidSpecs, err := getOrderedMandatoryMiddlewareSpecs(
+		h.ConfigRoot(),
+		h.MandatoryMiddlewareSpecs(),
+		endpointType.(string))
+	if err != nil {
+		return nil, errors.Wrap(
+			err, "Error getting ordered mandatory middleware specs",
+		)
+	}
+
+	return augmentEndpointSpec(espec, endpointConfigObj, midSpecs, mandatoryMidSpecs)
+}
+
+func getOrderedMandatoryMiddlewareSpecs(
+	cfgDir string,
+	middlewareSpecs map[string]*MiddlewareSpec,
+	classType string,
+) ([]MiddlewareSpec, error) {
+	middlewareObj := map[string][]string{}
+
+	const mandatoryMiddlewareFile = "middlewares/mandatory"
+	middlewareOrderingFile := filepath.Join(cfgDir, mandatoryMiddlewareFile+".yaml")
+	if _, err := os.Stat(middlewareOrderingFile); os.IsNotExist(err) {
+		// Cannot find yaml file, use json file instead
+		middlewareOrderingFile = filepath.Join(cfgDir, mandatoryMiddlewareFile+".json")
```

Comment:
The reason we check for .json file is that we previously used JSON files for config, and we need to have backward compatibility when we moved to yaml. Since this is a new feature and there is no backward compatibility requirement, I would say let's not support json at all. We will probably ditch support for json in the future.